File: GdPicture/WinForm/vb.net/DICOM Viewer/frmMain.vb
Option Strict On
Option Explicit On

Imports GdPicture14
Imports System.IO

Public Class frmMain

    Friend Function getNativeImage() As Integer
        Return m_ImageID
    End Function

    Private Enum OpenMode
        OpenModeUnknown = 0
        OpenModeFile = 1
        OpenModeFolder = 2
    End Enum

    Private Enum MouseMode
        MouseModeDefault = 0
        MouseModeWindowLeveling = 1
    End Enum

    Private m_GdPictureImaging As New GdPictureImaging
    Private m_ImageID As Integer
    Private m_openMode As OpenMode = OpenMode.OpenModeUnknown
    Private m_mouseMode As MouseMode = MouseMode.MouseModeWindowLeveling

    Private m_ww As Integer 'Current windows width
    Private m_wl As Integer 'Current windows level

    Private m_animating As Boolean
    Private WithEvents m_timer As New Timer
    Private m_stopwatch As New Stopwatch()
    Private m_TickCount As Long
    Private m_ImagesCount As Integer

    Private Sub frmMain_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim oLicenseManager As New GdPicture14.LicenseManager() 'Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
        oLicenseManager.RegisterKEY("XXXX") 'Please, replace XXXX by a valid demo or commercial license key.
        SplitContainer1.SplitterDistance = 620
        'If you need to set a different animation speed, this is the value you want to change (in milliseconds)
        'm_timer.Interval = 1000
    End Sub


    Private Sub OpenToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OpenToolStripMenuItem.Click
        CloseDocument()

        Dim fdl As New OpenFileDialog()
        fdl.Filter = "DICOM files (*.dcm)|*.dcm|All files (*.*)|*.*"
        If fdl.ShowDialog(Me) = Windows.Forms.DialogResult.OK Then
            Dim filePath As String = fdl.FileName
            If GdPictureDocumentUtilities.GetDocumentFormatFromFileName(filePath) = GdPicture14.DocumentFormat.DocumentFormatDICOM Then
                m_ImageID = m_GdPictureImaging.CreateGdPictureImageFromFile(filePath)
                If m_ImageID <> 0 Then
                    m_ImagesCount = Math.Max(m_GdPictureImaging.DicomGetPageCount(m_ImageID), 1)
                    m_openMode = OpenMode.OpenModeFile
                    GdViewer1.DisplayFromGdPictureImage(m_ImageID)
                    ThumbnailEx1.LoadFromGdPictureImage(m_ImageID)
                    GdViewer1.Focus()
                Else
                    MessageBox.Show("Unable to open the selected DICOM file, the error is: " + m_GdPictureImaging.GetStat().ToString(), "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1)
                End If
            Else
                MessageBox.Show("The selected file was not recognized as a DICOM file", "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1)
            End If
        End If
    End Sub


    Private Sub OpenDirectoryToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OpenDirectoryToolStripMenuItem.Click
        CloseDocument()

        Dim fdl As New FolderBrowserDialog
        If fdl.ShowDialog(Me) = Windows.Forms.DialogResult.OK Then
            Dim folderPath As String = fdl.SelectedPath
            Dim inputFiles As New List(Of String)
            Dim fifs As FileInfo() = New DirectoryInfo(folderPath).GetFiles()

            'list the names of all files in the specified directory
            For Each fif As IO.FileInfo In fifs
                Dim currentFile As String = fif.FullName
                If GdPictureDocumentUtilities.GetDocumentFormatFromFileName(currentFile) = GdPicture14.DocumentFormat.DocumentFormatDICOM Then
                    inputFiles.Add(currentFile)
                End If
            Next

            If inputFiles.Count > 0 Then
                m_ImagesCount = inputFiles.Count
                m_openMode = OpenMode.OpenModeFolder
                ThumbnailEx1.LoadFromFiles(inputFiles.ToArray())
                ThumbnailEx1.SelectItem(0)
                GdViewer1.Focus()
            Else
                MessageBox.Show("The selected folder does not includes any recognized DICOM file", "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1)
            End If

        End If
    End Sub


    Private Sub ThumbnailEx1_ItemSelectionChanged(ByVal Idx As Integer, ByVal Selected As Boolean) Handles ThumbnailEx1.ItemSelectionChanged
        If Selected Then
            Select Case m_openMode
                Case OpenMode.OpenModeFile
                    m_GdPictureImaging.DicomSelectPage(m_ImageID, Idx + 1)

                    If m_mouseMode = MouseMode.MouseModeWindowLeveling Then
                        If m_ww <> 0 And (m_ww <> m_GdPictureImaging.DicomGetDefaultWindowWidth(m_ImageID) Or m_wl <> m_GdPictureImaging.DicomGetDefaultWindowLevel(m_ImageID)) Then
                            m_GdPictureImaging.DicomSetWindowLevel(m_ImageID, m_ww, m_wl)
                        End If
                    End If

                    GdViewer1.Redraw()
                Case OpenMode.OpenModeFolder
                    Dim srcPath As String = ThumbnailEx1.GetItemFileSource(Idx)
                    GdViewer1.CloseDocument()
                    m_GdPictureImaging.ReleaseGdPictureImage(m_ImageID)
                    m_ImageID = m_GdPictureImaging.CreateGdPictureImageFromFile(srcPath)
                    If m_ImageID <> 0 Then
                        If m_mouseMode = MouseMode.MouseModeWindowLeveling Then
                            If m_ww <> 0 And (m_ww <> m_GdPictureImaging.DicomGetDefaultWindowWidth(m_ImageID) Or m_wl <> m_GdPictureImaging.DicomGetDefaultWindowLevel(m_ImageID)) Then
                                m_GdPictureImaging.DicomSetWindowLevel(m_ImageID, m_ww, m_wl)
                            End If
                        End If
                        GdViewer1.DisplayFromGdPictureImage(m_ImageID)
                    Else
                        'todo: handle error
                    End If
            End Select
        End If
    End Sub


    Private Sub CloseToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CloseToolStripMenuItem.Click
        CloseDocument()
    End Sub


    Private Sub ExitToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ExitToolStripMenuItem.Click
        Me.Close()
    End Sub


    Private Sub GotoFirstPageToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GotoFirstPageToolStripMenuItem.Click
        ThumbnailEx1.SelectItem(0)
    End Sub


    Private Function getSelectedItemIdx() As Integer
        For i As Integer = 0 To ThumbnailEx1.ItemCount - 1
            If ThumbnailEx1.GetItemSelectState(i) Then
                Return i
            End If
        Next

        Return -1
    End Function

    Private Sub displayFrame(ByVal frame As Integer)
        ThumbnailEx1.SelectItem(frame - 1)
    End Sub

    Private Sub displayNext()
        If m_ImagesCount > 1 Then
            Dim selectedIdx As Integer = getSelectedItemIdx()
            If selectedIdx + 1 < m_ImagesCount Then
                ThumbnailEx1.SelectItem(selectedIdx + 1)
            Else
                displayFrame(1)
            End If
        End If
    End Sub


    Private Sub displayPrevious()
        If m_ImagesCount > 1 Then
            ThumbnailEx1.SelectItem(getSelectedItemIdx() - 1)
        End If
    End Sub


    Private Sub GotoPreviousPageToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GotoPreviousPageToolStripMenuItem.Click
        ThumbnailEx1.SelectItem(getSelectedItemIdx() - 1)
    End Sub


    Private Sub GotoNextPageToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GotoNextPageToolStripMenuItem.Click
        ThumbnailEx1.SelectItem(getSelectedItemIdx() + 1)
    End Sub


    Private Sub GotoLastPageToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GotoLastPageToolStripMenuItem.Click
        ThumbnailEx1.SelectItem(ThumbnailEx1.ItemCount - 1)
    End Sub


    Private Sub Rotate90ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Rotate90ToolStripMenuItem.Click
        GdViewer1.Rotate(RotateFlipType.Rotate90FlipNone)
    End Sub


    Private Sub Rotate180ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Rotate180ToolStripMenuItem.Click
        GdViewer1.Rotate(RotateFlipType.Rotate180FlipNone)
    End Sub


    Private Sub Rotate270ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Rotate270ToolStripMenuItem.Click
        GdViewer1.Rotate(RotateFlipType.Rotate270FlipNone)
    End Sub


    Private Sub FlipXToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles FlipXToolStripMenuItem.Click
        GdViewer1.Rotate(RotateFlipType.RotateNoneFlipX)
    End Sub


    Private Sub CloseDocument()
        m_ww = 0
        m_wl = 0
        GdViewer1.CloseDocument()
        ThumbnailEx1.ClearAllItems()
        m_openMode = OpenMode.OpenModeUnknown
        Me.StatusStrip1.Text = ""
        If m_ImageID <> 0 Then
            m_GdPictureImaging.ReleaseGdPictureImage(m_ImageID)
            m_ImageID = 0
        End If
        m_ImagesCount = 0
    End Sub


    Private Sub ToolStripMenuItem10_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripMenuItem10.Click
        GdViewer1.PrintSetDocumentName("GdPicture Document Viewer Printing")
        GdViewer1.PrintSetAutoRotation(True)
        GdViewer1.PrintDialog(Me)
    End Sub


    Private Sub ToolStripMenuItem11_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripMenuProperties.Click
        Dim oForm As New frmProperties
        oForm.Show(Me)
    End Sub


    Private Sub Small64X64ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Small64X64ToolStripMenuItem.Click
        ThumbnailEx1.ThumbnailSize = New Size(64, 64)
        Small64X64ToolStripMenuItem.Checked = True
        Medium128X128ToolStripMenuItem.Checked = False
        Large256256ToolStripMenuItem.Checked = False
    End Sub


    Private Sub Medium128X128ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Medium128X128ToolStripMenuItem.Click
        ThumbnailEx1.ThumbnailSize = New Size(128, 128)
        Small64X64ToolStripMenuItem.Checked = False
        Medium128X128ToolStripMenuItem.Checked = True
        Large256256ToolStripMenuItem.Checked = False
    End Sub


    Private Sub Large256256ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Large256256ToolStripMenuItem.Click
        ThumbnailEx1.ThumbnailSize = New Size(256, 256)
        Small64X64ToolStripMenuItem.Checked = False
        Medium128X128ToolStripMenuItem.Checked = False
        Large256256ToolStripMenuItem.Checked = True
    End Sub


    Private Sub GdViewer1_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles GdViewer1.MouseMove
        If e.Button = Windows.Forms.MouseButtons.Left Then
            If m_mouseMode = MouseMode.MouseModeWindowLeveling Then

                Dim mouseX As Integer = Math.Max(0, Math.Min(e.X, GdViewer1.Width))
                Dim mouseY As Integer = Math.Max(0, Math.Min(e.Y, GdViewer1.Height))

                Dim newWW As Integer = Math.Max(CInt(m_GdPictureImaging.DicomGetMaxWindowWidth(m_ImageID) / GdViewer1.Width * mouseX), 1)
                Dim wlRange As Single = m_GdPictureImaging.DicomGetMaxWindowLevel(m_ImageID) - m_GdPictureImaging.DicomGetMinWindowLevel(m_ImageID)
                Dim newWL As Integer = CInt(m_GdPictureImaging.DicomGetMinWindowLevel(m_ImageID) + ((wlRange / GdViewer1.Height) * mouseY))
                If newWW <> m_ww Or newWL <> m_wl Then
                    m_ww = newWW
                    m_wl = newWL
                    If m_openMode = OpenMode.OpenModeFile Then
                        'trick: the thumbnail control can change the current used page no
                        m_GdPictureImaging.DicomSelectPage(m_ImageID, getSelectedItemIdx() + 1)
                    End If
                    If m_GdPictureImaging.DicomSetWindowLevel(m_ImageID, m_ww, m_wl) = GdPictureStatus.OK Then
                        GdViewer1.Redraw()
                    End If
                End If
            End If
        End If
    End Sub


    Private Sub GdViewer1_MouseWheel(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles GdViewer1.MouseWheel
        If e.Delta < 0 Then
            displayNext()
        Else
            displayPrevious()
        End If
    End Sub


    Private Sub GdViewer1_PageDisplayed() Handles GdViewer1.PageDisplayed
        'GdViewer1.Graphics.DrawString(Trim(Str(GdViewer1.CurrentPage)) & " / " & Trim(Str(GdViewer1.PageCount)), New Font("Arial", 20, Drawing.FontStyle.Bold, GraphicsUnit.Pixel), Brushes.Red, GdViewer1.Width - 100, 10)
        Dim WidthInches, HeightInches As Double

        If GdViewer1.GetDocumentType = DocumentType.DocumentTypePDF Then
            WidthInches = GdViewer1.PdfGetPageWidth / 72
            HeightInches = GdViewer1.PdfGetPageHeight / 72
        Else
            WidthInches = GdViewer1.PageWidth / GdViewer1.HorizontalResolution
            HeightInches = GdViewer1.PageHeight / GdViewer1.VerticalResolution
        End If

        Me.ToolStripStatusLabel1.Text = "Frame No : " + Trim(Str(GdViewer1.CurrentPage)) & " / " & Trim(Str(GdViewer1.PageCount)) + "   -   " +
                               "Zoom : " + Trim(Str(Math.Round(GdViewer1.Zoom * 100, 2))) & "%    -   " +
                               "Image Size (pixels) : " + Trim(Str(GdViewer1.PageWidth)) + " * " + Trim(Str(GdViewer1.PageHeight)) + "   -   " +
                               "Output Pixel Format : " + m_GdPictureImaging.GetPixelFormat(m_ImageID).ToString()
    End Sub


    Private Sub FlipYToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles FlipYToolStripMenuItem.Click
        GdViewer1.Rotate(RotateFlipType.RotateNoneFlipY)
    End Sub


    Private Sub GdViewer1_Paint(ByVal sender As Object, ByVal e As System.Windows.Forms.PaintEventArgs) Handles GdViewer1.Paint
        If m_ImageID <> 0 Then
            If m_mouseMode = MouseMode.MouseModeDefault Or m_ww = 0 Then
                m_ww = CInt(m_GdPictureImaging.DicomGetDefaultWindowWidth(m_ImageID))
                m_wl = CInt(m_GdPictureImaging.DicomGetDefaultWindowLevel(m_ImageID))
            End If

            Using f As New Font("Times New Roman", 20, Drawing.FontStyle.Bold, GraphicsUnit.Pixel)
                Using sf As New StringFormat
                    sf.Alignment = StringAlignment.Far
                    e.Graphics.DrawString("WL: " + m_wl.ToString() + " / WW: " + m_ww.ToString(), f, Brushes.Red, New RectangleF(0, 0, GdViewer1.Width - 10, 50), sf)
                End Using

            End Using
        End If
    End Sub


    Private Sub RestoreDefaultToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles RestoreDefaultToolStripMenuItem.Click
        m_ww = CInt(m_GdPictureImaging.DicomGetDefaultWindowWidth(m_ImageID))
        m_wl = CInt(m_GdPictureImaging.DicomGetDefaultWindowLevel(m_ImageID))
        If m_GdPictureImaging.DicomSetWindowLevel(m_ImageID, m_ww, m_wl) = GdPictureStatus.OK Then
            GdViewer1.Redraw()
        End If
    End Sub


    Private Sub EnableWindowLevelingToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EnableWindowLevelingToolStripMenuItem.Click
        EnableWindowLevelingToolStripMenuItem.Checked = Not EnableWindowLevelingToolStripMenuItem.Checked
        If EnableWindowLevelingToolStripMenuItem.Checked Then
            m_mouseMode = MouseMode.MouseModeWindowLeveling
        Else
            m_mouseMode = MouseMode.MouseModeDefault
        End If
    End Sub

    Private Sub AnimateToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles AnimateToolStripMenuItem.Click
        If m_animating Then
            endAnimation()
        Else
            startAnimation()
        End If
    End Sub

    Private Sub startAnimation()
        If Not m_animating AndAlso m_ImageID <> 0 Then
            GdViewer1.DisplayQuality = DisplayQuality.DisplayQualityLow
            resetFPS()
            m_TickCount = 0
            m_timer.Start()
            m_animating = True
        End If
    End Sub

    Private Sub endAnimation()
        If m_animating Then
            m_timer.[Stop]()
            m_animating = False
            GdViewer1.DisplayQuality = DisplayQuality.DisplayQualityBicubicHQ
        End If
    End Sub

    Private Sub resetFPS()
        m_TickCount = 0
        m_stopwatch.[Stop]()
        m_stopwatch.Start()
    End Sub

    Private Sub TimerEventProcessor(ByVal sender As Object, ByVal e As EventArgs) Handles m_timer.Tick
        displayNext()
        m_TickCount += 1
    End Sub

End Class

File: GdPicture/WinForm/C#/DICOM Viewer/frmMain.cs
using System.Drawing;
using System.Windows.Forms;
using System.Collections.Generic;
using System;
using GdPicture14;
using System.IO;

namespace GdPicture_DICOM_Viewer
{
    public partial class frmMain : Form
    {
        private enum OpenMode
        {
            OpenModeUnknown = 0,
            OpenModeFile = 1,
            OpenModeFolder = 2
        }

        private enum MouseMode
        {
            MouseModeDefault = 0,
            MouseModeWindowLeveling = 1
        }

        private readonly GdPictureImaging _gdPictureImaging;

        private int _nativeImageId;

        private OpenMode _openMode = OpenMode.OpenModeUnknown;

        private MouseMode _mouseMode = MouseMode.MouseModeWindowLeveling;

        private int _windowWidth;

        private int _windowLevel;

        private Timer _timer;

        private int _imagesCount;


        public frmMain()
        {
            _gdPictureImaging = new GdPictureImaging();
            InitializeComponent();
        }


        public int NativeImageId
        {
            get
            {
                return _nativeImageId;
            }
        }


        public void frmMain_Load(System.Object sender, System.EventArgs e)
        {
            LicenseManager licenseManager = new LicenseManager(); //Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
            licenseManager.RegisterKEY("XXXX"); //Please, replace XXXX by a valid demo or commercial license key.
            SplitContainer1.SplitterDistance = 620;
            _timer = new Timer();
            //If you need to set a different animation speed, this is the value you want to change (in milliseconds)
            //_timer.Interval = 1000;
            _timer.Tick += new EventHandler(DispatcherTimer_Tick);
        }


        public void OpenToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            CloseDocument();

            OpenFileDialog fdl = new OpenFileDialog();
            fdl.Filter = "DICOM files (*.dcm)|*.dcm|All files (*.*)|*.*";
            if (fdl.ShowDialog(this) == System.Windows.Forms.DialogResult.OK)
            {
                string filePath = fdl.FileName;
                if (GdPictureDocumentUtilities.GetDocumentFormat(filePath) == GdPicture14.DocumentFormat.DocumentFormatDICOM)
                {
                    _nativeImageId = _gdPictureImaging.CreateGdPictureImageFromFile(filePath);
                    if (_nativeImageId != 0)
                    {
                        _imagesCount = Math.Max(_gdPictureImaging.DicomGetPageCount(_nativeImageId), 1);
                        _openMode = OpenMode.OpenModeFile;
                        GdViewer1.DisplayFromGdPictureImage(_nativeImageId);
                        ThumbnailEx1.LoadFromGdPictureImage(_nativeImageId);
                        GdViewer1.Focus();
                    }
                    else
                    {
                        MessageBox.Show("Unable to open the selected DICOM file, the error is: " + _gdPictureImaging.GetStat(), "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1);
                    }
                }
                else
                {
                    MessageBox.Show("The selected file was not recognized as a DICOM file", "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1);
                }
            }
        }


        public void OpenDirectoryToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            CloseDocument();

            FolderBrowserDialog fdl = new FolderBrowserDialog();
            if (fdl.ShowDialog(this) == System.Windows.Forms.DialogResult.OK)
            {
                string folderPath = fdl.SelectedPath;
                List<string> inputFiles = new List<string>();
                FileInfo[] fifs = new DirectoryInfo(folderPath).GetFiles();

                //list the names of all files in the specified directory
                foreach (FileInfo fif in fifs)
                {
                    string currentFile = fif.FullName;
                    if (GdPictureDocumentUtilities.GetDocumentFormat(currentFile) == GdPicture14.DocumentFormat.DocumentFormatDICOM)
                    {
                        inputFiles.Add(currentFile);
                    }
                }

                if (inputFiles.Count > 0)
                {
                    _imagesCount = inputFiles.Count;
                    _openMode = OpenMode.OpenModeFolder;
                    ThumbnailEx1.LoadFromFiles(inputFiles.ToArray());
                    ThumbnailEx1.SelectItem(0);
                    GdViewer1.Focus();
                }
                else
                {
                    MessageBox.Show("The selected folder does not include any recognized DICOM file", "error", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1);
                }

            }
        }


        public void ThumbnailEx1_ItemSelectionChanged(int Idx, bool Selected)
        {
            if (Selected)
            {
                switch (_openMode)
                {
                    case OpenMode.OpenModeFile:
                        _gdPictureImaging.DicomSelectPage(_nativeImageId, Idx + 1);

                        if (_mouseMode == MouseMode.MouseModeWindowLeveling)
                        {
                            if (_windowWidth != 0 && (_windowWidth != _gdPictureImaging.DicomGetDefaultWindowWidth(_nativeImageId) || _windowLevel != _gdPictureImaging.DicomGetDefaultWindowLevel(_nativeImageId)))
                            {
                                _gdPictureImaging.DicomSetWindowLevel(_nativeImageId, _windowWidth, _windowLevel);
                            }
                        }
                        GdViewer1.Redraw();
                        break;
                    case OpenMode.OpenModeFolder:
                        string srcPath = ThumbnailEx1.GetItemFileSource(Idx);
                        GdViewer1.CloseDocument();
                        _gdPictureImaging.ReleaseGdPictureImage(_nativeImageId);
                        _nativeImageId = _gdPictureImaging.CreateGdPictureImageFromFile(srcPath);
                        if (_nativeImageId != 0)
                        {
                            if (_mouseMode == MouseMode.MouseModeWindowLeveling)
                            {
                                if (_windowWidth != 0 && (_windowWidth != _gdPictureImaging.DicomGetDefaultWindowWidth(_nativeImageId) || _windowLevel != _gdPictureImaging.DicomGetDefaultWindowLevel(_nativeImageId)))
                                {
                                    _gdPictureImaging.DicomSetWindowLevel(_nativeImageId, _windowWidth, _windowLevel);
                                }
                            }
                            GdViewer1.DisplayFromGdPictureImage(_nativeImageId);
                        }
                        else
                        {
                            MessageBox.Show("Image could not be loaded", _gdPictureImaging.GetStat().ToString(), MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1);
                        }
                        break;
                }
            }
        }


        public void CloseToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            CloseDocument();
        }


        public void ExitToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            Close();
        }


        public void GotoFirstPageToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.SelectItem(0);
        }


        private int GetSelectedItemIdx()
        {
            for (int i = 0; i <= ThumbnailEx1.ItemCount - 1; i++)
            {
                if (ThumbnailEx1.GetItemSelectState(i))
                {
                    return i;
                }
            }

            return -1;
        }


        private void DisplayFrame(int frame)
        {
            ThumbnailEx1.SelectItem(frame - 1);
        }


        private void DisplayNext()
        {
            if (_imagesCount > 1)
            {
                int selectedIdx = GetSelectedItemIdx();
                if (selectedIdx + 1 < _imagesCount)
                {
                    ThumbnailEx1.SelectItem(selectedIdx + 1);
                }
                else
                {
                    DisplayFrame(1);
                }
            }
        }


        private void DisplayPrevious()
        {
            if (_imagesCount > 1)
            {
                ThumbnailEx1.SelectItem(GetSelectedItemIdx() - 1);
            }
        }


        public void GotoPreviousPageToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.SelectItem(GetSelectedItemIdx() - 1);
        }


        public void GotoNextPageToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.SelectItem(GetSelectedItemIdx() + 1);
        }


        public void GotoLastPageToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.SelectItem(ThumbnailEx1.ItemCount - 1);
        }


        public void Rotate90ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.Rotate(RotateFlipType.Rotate90FlipNone);
        }


        public void Rotate180ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.Rotate(RotateFlipType.Rotate180FlipNone);
        }


        public void Rotate270ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.Rotate(RotateFlipType.Rotate270FlipNone);
        }


        public void FlipXToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.Rotate(RotateFlipType.RotateNoneFlipX);
        }


        private void CloseDocument()
        {
            _windowWidth = 0;
            _windowLevel = 0;
            GdViewer1.CloseDocument();
            ThumbnailEx1.ClearAllItems();
            _openMode = OpenMode.OpenModeUnknown;
            this.StatusStrip1.Text = "";
            if (_nativeImageId != 0)
            {
                _gdPictureImaging.ReleaseGdPictureImage(_nativeImageId);
                _nativeImageId = 0;
            }
            _imagesCount = 0;
        }


        public void ToolStripMenuItem10_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.PrintSetDocumentName("GdPicture Document Viewer Printing");
            GdViewer1.PrintDialog(this);
        }


        public void ToolStripMenuItem11_Click(System.Object sender, System.EventArgs e)
        {
            frmProperties oForm = new frmProperties();
            oForm.Show(this);
        }


        public void Small64X64ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.ThumbnailSize = new Size(64, 64);
            Small64X64ToolStripMenuItem.Checked = true;
            Medium128X128ToolStripMenuItem.Checked = false;
            Large256256ToolStripMenuItem.Checked = false;
        }


        public void Medium128X128ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.ThumbnailSize = new Size(128, 128);
            Small64X64ToolStripMenuItem.Checked = false;
            Medium128X128ToolStripMenuItem.Checked = true;
            Large256256ToolStripMenuItem.Checked = false;
        }


        public void Large256256ToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            ThumbnailEx1.ThumbnailSize = new Size(256, 256);
            Small64X64ToolStripMenuItem.Checked = false;
            Medium128X128ToolStripMenuItem.Checked = false;
            Large256256ToolStripMenuItem.Checked = true;
        }


        public void GdViewer1_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (e.Button == System.Windows.Forms.MouseButtons.Left)
            {
                if (_mouseMode == MouseMode.MouseModeWindowLeveling)
                {

                    int mouseX = Math.Max(0, Math.Min(e.X, GdViewer1.Width));
                    int mouseY = Math.Max(0, Math.Min(e.Y, GdViewer1.Height));

                    int newWW = Math.Max((int)(_gdPictureImaging.DicomGetMaxWindowWidth(_nativeImageId) / GdViewer1.Width * mouseX), 1);
                    float wlRange = _gdPictureImaging.DicomGetMaxWindowLevel(_nativeImageId) - _gdPictureImaging.DicomGetMinWindowLevel(_nativeImageId);
                    int newWL = (int)(_gdPictureImaging.DicomGetMinWindowLevel(_nativeImageId) + ((wlRange / GdViewer1.Height) * mouseY));
                    if (newWW != _windowWidth | newWL != _windowLevel)
                    {
                        _windowWidth = newWW;
                        _windowLevel = newWL;
                        if (_openMode == OpenMode.OpenModeFile)
                        {
                            //trick: the thumbnail control can change the current used page no
                            _gdPictureImaging.DicomSelectPage(_nativeImageId, GetSelectedItemIdx() + 1);
                        }
                        if (_gdPictureImaging.DicomSetWindowLevel(_nativeImageId, _windowWidth, _windowLevel) == GdPictureStatus.OK)
                        {
                            GdViewer1.Redraw();
                        }
                    }
                }
            }
        }


        public void GdViewer1_MouseWheel(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (e.Delta < 0)
            {
                DisplayNext();
            }
            else
            {
                DisplayPrevious();
            }
        }


        public void GdViewer1_PageDisplayed()
        {
            ToolStripStatusLabel1.Text = "Frame No : " + GdViewer1.CurrentPage.ToString().Trim() + " / " + GdViewer1.PageCount.ToString().Trim() + "   -   " + "Zoom : " +
                Math.Round(GdViewer1.Zoom * 100, 2).ToString().Trim() + ("%    -   " + "Image Size (pixels) : " + GdViewer1.PageWidth.ToString().Trim() + " * " +
                GdViewer1.PageHeight.ToString().Trim() + "   -   " + "Output Pixel Format : " + _gdPictureImaging.GetPixelFormat(_nativeImageId));
        }


        public void FlipYToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            GdViewer1.Rotate(RotateFlipType.RotateNoneFlipY);
        }


        public void GdViewer1_Paint(object sender, System.Windows.Forms.PaintEventArgs e)
        {
            if (_nativeImageId != 0)
            {
                if (_mouseMode == MouseMode.MouseModeDefault | _windowWidth == 0)
                {
                    _windowWidth = (int)(_gdPictureImaging.DicomGetDefaultWindowWidth(_nativeImageId));
                    _windowLevel = (int)(_gdPictureImaging.DicomGetDefaultWindowLevel(_nativeImageId));
                }

                using (Font f = new Font("Times New Roman", 20, System.Drawing.FontStyle.Bold, GraphicsUnit.Pixel))
                {
                    using (StringFormat sf = new StringFormat())
                    {
                        sf.Alignment = StringAlignment.Far;
                        e.Graphics.DrawString("WL: " + _windowLevel + " / WW: " + _windowWidth, f, Brushes.Red, new RectangleF(0, 0, GdViewer1.Width - 10, 50), sf);
                    }
                }
            }
        }


        public void RestoreDefaultToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            _windowWidth = (int)(_gdPictureImaging.DicomGetDefaultWindowWidth(_nativeImageId));
            _windowLevel = (int)(_gdPictureImaging.DicomGetDefaultWindowLevel(_nativeImageId));
            if (_gdPictureImaging.DicomSetWindowLevel(_nativeImageId, _windowWidth, _windowLevel) == GdPictureStatus.OK)
            {
                GdViewer1.Redraw();
            }
        }


        public void EnableWindowLevelingToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
        {
            EnableWindowLevelingToolStripMenuItem.Checked = !EnableWindowLevelingToolStripMenuItem.Checked;
            if (EnableWindowLevelingToolStripMenuItem.Checked)
            {
                _mouseMode = MouseMode.MouseModeWindowLeveling;
            }
            else
            {
                _mouseMode = MouseMode.MouseModeDefault;
            }
        }


        private void DispatcherTimer_Tick(object sender, EventArgs e)
        {
            DisplayNext();
        }
    }
}
